### PR TITLE
Class based views

### LINKED ISSUE #24: Rewrite views using Class-Based Views
1. Read about class-based views in docs
- [ ] intro https://docs.djangoproject.com/en/1.10/topics/class-based-views/intro/
- [ ] advanced https://docs.djangoproject.com/en/1.10/topics/class-based-views/
2. Rewrite yours

p.s - don't forget that tests MUST be done before you start this task

### PR DESCRIPTION
**Pull request by makkaronis4e**
Add class based views
Resolve #24

### CODE REVIEW

**Comment by makkaronis4e** [comment]
@webknjaz done

File: blog/views.py [hunk 1]
```diff
@@ -1,53 +1,67 @@
...
+    """Protect views that need to show only for authorised users."""
+
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Rewrite standart method and it to decorator."""
+        return super().dispatch(*args, **kwargs)
+
 
+class PostList(ListView):
+    """Show all your Post objects."""
```
Comment (webknjaz) [inline blog/views.py]: Your? Whom is it talking about?

File: blog/views.py [hunk 2]
```diff
@@ -1,53 +1,67 @@
...
+from django.views.generic import DetailView, ListView, UpdateView, View, FormView
+
+from .forms import PostForm
+from .models import Post
+
+
+class Protected(View):
+    """Protect views that need to show only for authorised users."""
+
+    @method_decorator(login_required)
```
Comment (webknjaz) [inline blog/views.py:17]: Did it work like this before rewrite?

File: blog/views.py [hunk 3]
```diff
@@ -1,53 +1,67 @@
...
-@login_required
-def post_new(request):
-    """Render new post in blog."""
-    if request.method == 'POST':
+    model = Post
+    template_name = 'blog/post_detail.html'
+
+
+class NewPost(FormView, Protected):
+    """Return page for adding new Post."""
```
Comment (webknjaz) [inline blog/views.py]: Don't use actionable verbs in class docstring as it describes some entity, not an action.

File: blog/views.py [hunk 4]
```diff
@@ -1,53 +1,67 @@
...
+    template_name = 'blog/post_edit.html'
+
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
+        return super().form_valid(form)
+
+    def post(self, request, *args, **kwargs):
+        """Get info from form and save it as 'post' object."""
         form = PostForm(request.POST)
```
Comment (webknjaz) [inline blog/views.py]: You shouldn't need to initialize your own form instance. It should be pre-populated by internal CBV machinery and available as `self.form`

File: blog/views.py [hunk 5]
```diff
@@ -1,53 +1,67 @@
...
         if form.is_valid():
```
Comment (webknjaz) [inline blog/views.py]: `post()` method will only be called if it's valid (based on the `form_valid()` method call earlier), so this check is also useless.

File: blog/views.py [hunk 6]
```diff
@@ -1,53 +1,67 @@
...
+
+    form_class = PostForm
+    template_name = 'blog/post_edit.html'
+
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
+        return super().form_valid(form)
+
+    def post(self, request, *args, **kwargs):
```
Comment (webknjaz) [inline blog/views.py]: I bet `post()` method does not accept request as an argument. args/kwargs also look unrealistic

File: blog/views.py [hunk 7]
```diff
@@ -1,53 +1,67 @@
...
+        """Get info from form and save it as 'post' object."""
         form = PostForm(request.POST)
         if form.is_valid():
             post = form.save(commit=False)
             post.author = request.user
```
Comment (webknjaz) [inline blog/views.py]: You already do this in `form_valid()` method

File: blog/views.py [hunk 8]
```diff
@@ -1,53 +1,67 @@
...
-            post.published_date = timezone.now()
             post.save()
             return redirect('post_detail', pk=post.pk)
```
Comment (webknjaz) [inline blog/views.py]: Based on above, I'd say that it is never called. success_url is used instead.

File: blog/views.py [hunk 9]
```diff
@@ -1,53 +1,67 @@
...
+        return super().form_valid(form)
+
+    def post(self, request, *args, **kwargs):
+        """Get info from form and save it as 'post' object."""
         form = PostForm(request.POST)
         if form.is_valid():
             post = form.save(commit=False)
             post.author = request.user
-            post.published_date = timezone.now()
             post.save()
```
Comment (webknjaz) [inline blog/views.py]: And this of course is done by CBV as well

File: blog/views.py [hunk 10]
```diff
@@ -1,53 +1,67 @@
...
-            return redirect('post_detail', pk=post.pk)
-    else:
-        form = PostForm(instance=post)
-    return render(request, 'blog/post_edit.html', {'form': form})
+class EditPost(UpdateView, Protected):
+    """View is for editing post."""
+
+    model = Post
+    fields = ['title', 'text']
+    success_url = reverse_lazy('post_list')
```
Comment (webknjaz) [inline blog/views.py]: You'd better redirect to post

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: blog/views.py (hunk 6 above)
Comment (webknjaz) [reply blog/views.py]: So remove it.

**Review by webknjaz — CHANGES_REQUESTED** [review]

**Review by webknjaz — APPROVED** [review]

**Pull request merged**